### PR TITLE
[incidentist] Use incident search API from Datadog API client

### PR DESCRIPTION
**Pull request by kkhor-datadog**
### Changes

This PR addresses two separate, but related issues:
- After exposing the report generation logic in a new package created in https://github.com/xornivore/incidentist/pull/7, various types that are only used internally are now exported as well: https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L29 https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L148 https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L154 https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L162 The exported types show up in the package documentation on https://pkg.go.dev/github.com/xornivore/incidentist as well. This did not previously occur as all of these exported types were in the `main` package, which is unimportable.
- After attempting to use the new `report` package in a different Go project that was using a different version of `datadog-api-client-go`, we encountered errors such as the following:
```
/go/pkg/mod/github.com/xornivore/incidentist@v0.0.0-20240311145355-33efc1adfbc0/report/search_incidents.go:321:178: too many arguments in call to client.PrepareRequest
       	have (context.Context, string, string, interface{}, map[string]string, url.Values, url.Values, string, string, []byte)
       	want (context.Context, string, string, interface{}, map[string]string, url.Values, url.Values, *datadog.FormFile)
```

The common thread between these issues is that incidentist has its own implementation of an API to query the Datadog incident search endpoint `/api/v2/incidents/search`: https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L237-L242

As a result of this bespoke incident search API implementation: 
- Attempting to use the new `report` package in a different Go project that was using a different, non-backward compatible version of `datadog-api-client-go` led to errors. 
- The response types used for unmarshaling the data received from the incident search API were publicly exported, even though to a large extent, the response types largely mirrored the ones currently available in the Datadog API

This made sense when incidentist was originally implemented against Datadog API client version 1.6, as the Datadog API client did not have functionality for querying this incident search endpoint yet. However, SDK support for querying this API has since been added in v2.7.0: https://github.com/DataDog/datadog-api-client-go/pull/1769

This PR:
- Updates incidentist to use the incident search API from the Datadog API client, instead of rolling its own implementation. As a result, all of the exported types previously only used for the API implementation are eliminated
- Updates dependency versions in `go.mod` as needed
- Fleshes out documentation of the `GenerateRequest` struct, as this stores report request parameters that external package users use to interact with the report API

### Testing
Tested by generating a report from the existing code and the new code, and verified that there were no diffs, as expected.

Note that there is now a warning log about the use of the incident search API, which is technically still in public beta: https://docs.datadoghq.com/api/latest/incidents/#search-for-incidents
```
2024/03/11 17:24:22 WARNING: Using unstable operation 'v2.SearchIncidents'
```

### CODE REVIEW

File: report/search_incidents.go [hunk 1]
```diff
@@ -1,397 +1,39 @@
...
-	}
-	return errors.New("undefined response type")
-}
-
 type searchRequest struct {
 	createdAfter  *int64
 	createdBefore *int64
 	tags          []string
-	pageSize      *int
-	pageOffset    *int
```
Comment (kkhor-datadog) [inline report/search_incidents.go:144]: These were never populated, even in the original API implementation: https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L421-L427

File: report/search_incidents.go [hunk 2]
```diff
@@ -1,397 +1,39 @@
...
 	var queryOpts []string
-
 	if r.createdBefore != nil {
 		queryOpts = append(queryOpts, fmt.Sprintf("created_before:%d", *r.createdBefore))
 	}
 	if r.createdAfter != nil {
 		queryOpts = append(queryOpts, fmt.Sprintf("created_after:%d", *r.createdAfter))
 	}
 	queryOpts = append(queryOpts, r.tags...)
+	query := strings.Join(queryOpts, " ")
```
Comment (kkhor-datadog) [inline report/search_incidents.go:31]: This matches the previous implementation constructing the query: https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L258

File: report/search_incidents.go [hunk 3]
```diff
@@ -1,397 +1,39 @@
...
-	req, err := client.PrepareRequest(ctx, localVarPath, localVarHTTPMethod, localVarPostBody, localVarHeaderParams, localVarQueryParams, localVarFormParams, localVarFormFileName, localVarFileName, localVarFileBytes)
-	if err != nil {
-		return localVarReturnValue, nil, err
-	}
-
-	localVarHTTPResponse, err := client.CallAPI(req)
-	if err != nil || localVarHTTPResponse == nil {
-		return localVarReturnValue, localVarHTTPResponse, err
-	}
+	queryParams := datadogV2.NewSearchIncidentsOptionalParameters().WithSort(datadogV2.INCIDENTSEARCHSORTORDER_CREATED_ASCENDING)
```
Comment (kkhor-datadog) [inline report/search_incidents.go:33]: This matches the sort order specified by the previous query: https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L270

File: report/search_incidents.go [hunk 4]
```diff
@@ -430,37 +72,46 @@ func fetchIncidents(team, ddApiKey, ddAppKey string, since, until time.Time) ([]
...
-		return nil, fmt.Errorf("Error when calling `IncidentsApi.SearchIncidents`: %w", err)
+		return nil, fmt.Errorf("Error when searching for incidents: %w", err)
 	}
 
-	if resp.Incidents == nil {
+	if resp.Data.Attributes == nil || resp.Data.Attributes.Incidents == nil {
 		return nil, nil
 	}
 
+	// The raw API response actually contains the incident commander embedded in the incidents, but the SDK doesn't expose it, as this is technically not JSON:API compliant. The SDK only exposes an ID in the relationships.
```
Comment (kkhor-datadog) [inline report/search_incidents.go:82]: The previous implementation was parsing the `commander` field, so it was relying on this technically non-compliant response format: https://github.com/xornivore/incidentist/blob/33efc1adfbc0599a810e5e618c26307ee03449ba/report/search_incidents.go#L170-L171

**Review by xornivore — APPROVED** [review]
Thank you for removing the hackery and using the official API method 😄

**Pull request merged**